File: py/structure/tree/binary_tree/traversal/PreOrderTraversal.py
#!/usr/bin/env python
# -*- coding:gbk -*-

from typing import List

class TreeNode:
    def __init__(self, value: int):
        self.value, self.left, self.right = value, None, None

class PreOrderTraversal:
    def traversal(self, root: TreeNode) -> List[int]:
        result, stack, current = [], [], root
        while current or stack:
            while current:
                result.append(current.value)

                stack.append(current)
                current = current.left
            current = stack.pop().right
        return result

File: cpp/structure/tree/binary_tree/traversal/PreOrderTraversal.cpp
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

struct TreeNode {
    int value;
    TreeNode *left;
    TreeNode *right;

    TreeNode(int value) : value(value), left(nullptr), right(nullptr) {}
};

class PreOrderTraversal {
public:
    vector<int> traversal(TreeNode *root) {
        vector<int> result(0);
        stack<TreeNode *> stack;
        TreeNode *current = root;
        while (current || !stack.empty()) {
            while (current) {
                result.emplace_back(current->value);

                stack.push(current);
                current = current->left;
            }
            current = stack.top()->right;
            stack.pop();
        }
        return result;
    }
};
